### DISCUSSION TITLE
Program Blocked.

### DISCUSSION THREAD

**Discussion by CommanderBash** [Kistler-Group/sdbus-cpp#431] [General]
When inside dbus method call in one thread and other thread accessing createSignal function it locks for ever.
Is this normal or this a bug? I can provide sample code if needed.  Method call and signal are registered on the same IObject

`T1` Recives a method call. Method call needs `T2` to shut down so it waits the thread to join.
`T2` Tries to send out a signal onto dbus, but will lock inside createSignal. so It will never exit.

**Comment by sangelovic** [comment]
This is a bug, but not of sdbus-c++, but the program using sdbus-c++. The two mentioned concurrent accesses to the D-Bus connection are nicely handled (serialized) by sdbus-c++, but the program creates a deadlock situation by having one depend on the other, of which sdbus-c++ has no idea.

One way to solve that off the top of my head would be to use server-side async method. The method callback finishes while the async result instance is stored somewhere. The underlying Connection resource is free for `createSignal`. Once signal is created (and sent out) in T2, the program can write to the async result instance, effectively informing the waiting caller of the finished function call.

**Reply by CommanderBash** [reply]
Just to clarify its an incoming method call. If that changes anything.

**Reply by sangelovic** [reply]
Yes, that's how I understood it. So no change.
